pkg/storage/utils/eosfs/eosfs.go

```diff
@@ -519,42 +529,56 @@ func (fs *eosfs) AddGrant(ctx context.Context, ref *provider.Reference, g *provi
 		return errors.Wrap(err, "eosfs: no user in ctx")
 	}
 
+	uid, gid, err := fs.getUserUIDAndGID(ctx, u)
+	if err != nil {
+		return err
+	}
+
 	p, err := fs.resolve(ctx, u, ref)
 	if err != nil {
 		return errors.Wrap(err, "eosfs: error resolving reference")
 	}
 
 	fn := fs.wrap(ctx, p)
 
-	uid, gid, err := fs.getUserUIDAndGID(ctx, u)
-	if err != nil {
-		return err
+	// workaround until EOS has native support
+	// for persisting order of the acl entries
+	if grants.IsDenial(g) {
+		gs, err := fs.ListGrants(ctx, ref)
+		if err != nil {
+			return err
+		}
+
+		// update list of grants
+		grants.AddGrant(&gs, g)
+
+		acls, err := fs.getEosACLs(ctx, gs)
+		if err != nil {
+			return err
+		}
+
+		if err = fs.setACLs(ctx, fs.c, uid, gid, fn, acls); err != nil {
+			return errors.Wrap(err, "eosfs: error setting acls")
+		}
+		return nil
+
 	}
 
-	gs, err := fs.ListGrants(ctx, ref)
+	eosACL, err := fs.getEosACL(ctx, g)
 	if err != nil {
 		return err
 	}
 
-	// update list of grants
-	grants.AddGrant(&gs, g)
-
-	acls, err := fs.getEosACLs(ctx, gs)
+	rootUID, rootGID, err := fs.getRootUIDAndGID(ctx)
 	if err != nil {
 		return err
 	}
 
-	// rootUID, rootGID, err := fs.getRootUIDAndGID(ctx)
-
-	if err = setACLs(ctx, fs.c, uid, gid, fn, acls); err != nil {
-		return errors.Wrap(err, "eosfs: error setting acls")
+	err = fs.c.AddACL(ctx, uid, gid, rootUID, rootGID, fn, eosACL)
```

Comment:
I think we should continue to use setACLs function as for denial grants, since for now when you add a new acl with the acl add command, the acls are shuffled. We risk that the denials, that we decided to keep at the end, are randomly put before a non denial acl.